### ISSUE TITLE
DESTROY, message not in perldiag pod

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6970]
Migrated from [rt.perl.org#24627](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24627) (status was 'resolved')

Searchable as RT24627$

**Comment by p5pRT** [comment]
### From ambrus@math.bme.hu

#### Created by ambrus@king.diebold.hu

This is a bug report for perl from ambrus@&#8203;king\.diebold\.hu\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.1\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

DESTROY error message missing from perldiag pod

There is a perl core error message missing from perldiags\.  The error is&#8203;:
DESTROY created new reference to dead object \.\.\.

Here are two examples of the message&#8203;: the command line to run perl\-5\.8\.1\,
and the diagnostics it emits \(m/^\(\\d\*\)\.\*?> / is my prompt\, $2 is the
errorlevel \($?\)\)&#8203;:

am ~/a/perl\-5\.8\.1> perl \-we'sub DESTROY\{@&#8203;b=@&#8203;\_;warn\};bless\[\];'
Name "main&#8203;::b" used only once&#8203;: possible typo at \-e line 1\.
Warning&#8203;: something's wrong at \-e line 1\.
Warning&#8203;: something's wrong at \-e line 1 during global destruction\.
DESTROY created new reference to dead object 'main' during global
destruction\.
255 am ~/a/perl\-5\.8\.1> perl \-we'sub DESTROY\{@&#8203;b=@&#8203;\_;warn\};bless\[\];bless\[\];'
Name "main&#8203;::b" used only once&#8203;: possible typo at \-e line 1\.
Warning&#8203;: something's wrong at \-e line 1\.
Warning&#8203;: something's wrong at \-e line 1\.
Warning&#8203;: something's wrong at \-e line 1\.
Warning&#8203;: something's wrong at \-e line 1 during global destruction\.
&nbsp;        \(in cleanup\) DESTROY created new reference to dead object 'main' at \-e line 1 during global destruction\.
&nbsp;        \(in cleanup\) DESTROY created new reference to dead object 'main' at \-e line 1 during global destruction\.
&nbsp;        \.\.\.caught at \-e line 1 during global destruction\.
DESTROY created new reference to dead object 'main' during global destruction\.
255 am ~/a/perl\-5\.8\.1>

The DESTROY sub I've given is of course wrong\, as it really creates a
new reference\, when perl should destroy all objects\.  However\, this
error is generated by the perl core\, more exactly the Perl\_sv\_clear
function def'd in "sv\.c"\.  I belive that every such message should be
documented in the perldiag pod file\, but this one isn't\.  Please write
a documentation for it\.

Also\, note the suffix "during global destruction\."\, which is added after
the line number each time perl formats an error message\.  See function
Perl\_vmess in "util\.c"\.  This suffix is mentioned only in perlobj\.pod\,
but maybe is should be documented in perldiag\.pod too\.

Please document these error messages\.

Thanks in advance\,
Ambrus\, \<ambrus@&#8203;math\.bme\.hu>

<details><summary>Perl Info</summary>

```

Flags:
    category=docs
    severity=low

Site configuration information for perl v5.8.1:

Configured by ambrus at Mon Sep 29 12:32:01 CEST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.21, archname=i686-linux
    uname='linux king 2.4.21 #4 fri sep 19 17:51:07 cest 2003 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=unde    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SO
    optimize='-O3',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldbm -ldb -ldl -lm -lcrypt -lutil -lc -lposix
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc -lposix
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.1.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:



@INC for perl v5.8.1:
    /usr/local/lib/perl5/5.8.1/i686-linux
    /usr/local/lib/perl5/5.8.1
    /usr/local/lib/perl5/site_perl/5.8.1/i686-linux
    /usr/local/lib/perl5/site_perl/5.8.1
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.1:
    HOME=/home/ambrus
    LANG=POSIX
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/bin:/bin:/usr/bin:~/scripts:/usr/games/bin:/usr/X11R6/bin:/o    PERL_BADLANG (unset)
    PERL_BADLANG (unset)
    SHELL=/bin/bash




```
</details>

**Comment by p5pRT** [comment]
### From tassilo.parseval@post.rwth-aachen.de

On Mon\, Dec 08\, 2003 at 06&#8203;:29&#8203;:43PM \+0000 Zsban Ambrus wrote&#8203;:

> DESTROY error message missing from perldiag pod
> 
> There is a perl core error message missing from perldiags\.  The error is&#8203;:
> DESTROY created new reference to dead object \.\.\.
> 
> Here are two examples of the message&#8203;: the command line to run perl\-5\.8\.1\,
> and the diagnostics it emits \(m/^\(\\d\*\)\.\*?> / is my prompt\, $2 is the
> errorlevel \($?\)\)&#8203;:
> 
> am ~/a/perl\-5\.8\.1> perl \-we'sub DESTROY\{@&#8203;b=@&#8203;\_;warn\};bless\[\];'
> Name "main&#8203;::b" used only once&#8203;: possible typo at \-e line 1\.
> Warning&#8203;: something's wrong at \-e line 1\.
> Warning&#8203;: something's wrong at \-e line 1 during global destruction\.
> DESTROY created new reference to dead object 'main' during global
> destruction\.
> 255 am ~/a/perl\-5\.8\.1> perl \-we'sub DESTROY\{@&#8203;b=@&#8203;\_;warn\};bless\[\];bless\[\];'
> Name "main&#8203;::b" used only once&#8203;: possible typo at \-e line 1\.
> Warning&#8203;: something's wrong at \-e line 1\.
> Warning&#8203;: something's wrong at \-e line 1\.
> Warning&#8203;: something's wrong at \-e line 1\.
> Warning&#8203;: something's wrong at \-e line 1 during global destruction\.
>         \(in cleanup\) DESTROY created new reference to dead object 'main' at \-e line 1 during global destruction\.
>         \(in cleanup\) DESTROY created new reference to dead object 'main' at \-e line 1 during global destruction\.
>         \.\.\.caught at \-e line 1 during global destruction\.
> DESTROY created new reference to dead object 'main' during global destruction\.
> 255 am ~/a/perl\-5\.8\.1>

With a slight variation on the above schemes\, it gets worse&#8203;:

&nbsp;    sub DESTROY \{ $b = \\$\_\[0\] \}
&nbsp;    END \{ print $$b \}
&nbsp;    bless \[\];
&nbsp;    \_\_END\_\_
&nbsp;    \<unpastable garbage> €š¬°ŽžŒÀÄÈÌÐÔØÜàäèìðôøü=UNKNOWN\(0x812b620\)

Also note that perl wont complain about creating a new reference to the
dead object when assigning a reference to $\_\[0\] to a global variable\. 

Tassilo
\-\- 
$\_=q\#"\,\}\]\)\!JAPH\!qq\(tsuJ\[\{@&#8203;"tnirp\}3\.\.0\}\_$;//&#8203;::niam/s~=\)\]3\[\)\)\_$\-3\(rellac\(=\_$\(\{
pam\{rekcahbus\}\)\(rekcah\{lrePbus\}\)\(lreP\{rehtonabus\}\)\!JAPH\!qq\(rehtona\{tsuJbus\#;
$\_=reverse\,s\+\(?\<=sub\)\.\+q\#q\!'"qq\.\\t$&\."'\!\#\+sexisexiixesixeseg;y~\\n~~dddd;eval

**Comment by p5pRT** [comment]
### From @rgs

Tassilo von Parseval wrote&#8203;:
> On Mon\, Dec 08\, 2003 at 06&#8203;:29&#8203;:43PM \+0000 Zsban Ambrus wrote&#8203;:
> 
> > DESTROY error message missing from perldiag pod

I've added it as change \#21877 :

==== //depot/perl/pod/perldiag\.pod\#365 \(text\) ====

@&#8203;@&#8203; \-1285\,6 \+1285\,12 @&#8203;@&#8203;
 long for Perl to handle\.  You have to be seriously twisted to write code
 that triggers this error\.
 
\+=item DESTROY created new reference to dead object '%s'
\+
\+\(F\) A DESTROY\(\) method created a new reference to the object which is
\+just being DESTROYed\. Perl is confused\, and prefers to abort rather than
\+to create a dangling reference\.
\+
 =item Did not produce a valid header
 
 See Server error\.

> 
> With a slight variation on the above schemes\, it gets worse&#8203;:
> 
>     sub DESTROY \{ $b = \\$\_\[0\] \}
>     END \{ print $$b \}
>     bless \[\];
>     \_\_END\_\_
>     \<unpastable garbage> \_\_\_\_\_¬°\_\_\_\_\_\_ÀÄÈÌÐÔØÜàäèìðôøü=UNKNOWN\(0x812b620\)
> 
> Also note that perl wont complain about creating a new reference to the
> dead object when assigning a reference to $\_\[0\] to a global variable\. 

Well\, that's a more serious bug \(and I would think that creating a new ticket
for it may be worthwhile\)

**Comment by p5pRT** [comment]
### From tassilo.parseval@post.rwth-aachen.de

On Fri\, Dec 12\, 2003 at 01&#8203;:08&#8203;:32AM \+0100 Rafael Garcia\-Suarez wrote&#8203;:

> Tassilo von Parseval wrote&#8203;:

> > With a slight variation on the above schemes\, it gets worse&#8203;:
> > 
> >     sub DESTROY \{ $b = \\$\_\[0\] \}
> >     END \{ print $$b \}
> >     bless \[\];
> >     \_\_END\_\_
> >     \<unpastable garbage> \_\_\_\_\_¬°\_\_\_\_\_\_ÀÄÈÌÐÔØÜàäèìðôøü=UNKNOWN\(0x812b620\)
> > 
> > Also note that perl wont complain about creating a new reference to the
> > dead object when assigning a reference to $\_\[0\] to a global variable\. 
> 
> Well\, that's a more serious bug \(and I would think that creating a new ticket
> for it may be worthwhile\)

Good news\, everyone\. The above was produced with 5\.8\.0\, but beginning
with 5\.8\.1 this problem no longer exists&#8203;:

ethan@&#8203;ethan&#8203;:\~$ perl5\.8\.1 \-e 'sub DESTROY \{ $b = \\$\_\[0\] \} END \{ print $$b \} bless \[\];'
main=ARRAY\(0x8136420\)DESTROY created new reference to dead object 'main'
during global destruction\.

Tassilo
\-\- 
$\_=q\#"\,\}\]\)\!JAPH\!qq\(tsuJ\[\{@&#8203;"tnirp\}3\.\.0\}\_$;//&#8203;::niam/s~=\)\]3\[\)\)\_$\-3\(rellac\(=\_$\(\{
pam\{rekcahbus\}\)\(rekcah\{lrePbus\}\)\(lreP\{rehtonabus\}\)\!JAPH\!qq\(rehtona\{tsuJbus\#;
$\_=reverse\,s\+\(?\<=sub\)\.\+q\#q\!'"qq\.\\t$&\."'\!\#\+sexisexiixesixeseg;y~\\n~~dddd;eval

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
